Added quarantine feature. Ported from Twister V1

gchwier opened the pull request:
Fixes #23 

  * added args to pytest (`--quarantine-list=FILENAME` , to verify use `-m quarantine`)
  * quarantine implementation added as a separate plugin
  * allowed multiple yaml files with quarantine lists
  * unit/component tests
  * quarantine yaml verification



Signed-off-by: Grzegorz Chwierut <grzegorz.chwierut@nordicsemi.no>

fundakol commented:
Would it work with pytest test?

For example If I have such test:
```python
# FILE: tests/foo_test.py
import pytest

@pytest.mark.platform('native_posix')
def test_foo():
    assert True
```
and quarantine file like:
```yaml
- scenarios:
    - tests.foo_test::test_foo
  platforms:
    - native_posix
```

gchwier commented:
Thanks for reply. It can probably work in such a way; I will check it.

I rather thought about adding
`pytest.mark.skip('reason')`
 to quarantined tests, like it is done in `yaml_test_function.py`. Then after running tests, we can find those tests in the output log:
`tests/kernel/common/testcase.yaml::kernel.common[native_posix] <- ../../upstream_workspace/twister/src/twister2/yaml_test_function.py SKIPPED (Skipped in yaml specification) [ 50%]`

I will update this PR soon.

fundakol commented:
You have right with pytest.mark.skip('reason'), it makes more sense for pytest tests and it is simpler to use. I guess quarantine is only way to skip tests from yaml file.

gchwier commented:
Added unittest:

```
$ pytest tests -k quarantine -v
============================================ test session starts =============================================
platform linux -- Python 3.8.10, pytest-7.2.0, pluggy-1.0.0 -- /home/grch/pyenvs/twister/bin/python3
cachedir: .pytest_cache
rootdir: /home/grch/upstream_workspace/twister, configfile: tox.ini
plugins: subtests-0.9.0, cov-4.0.0
collected 60 items / 53 deselected / 7 selected                                                              

tests/twister2_plugin_quarantine_test.py::test_if_pytest_use_quarantine_file PASSED                    [ 14%]
tests/twister2_plugin_quarantine_test.py::test_if_pytest_use_two_quarantine_files PASSED               [ 28%]
tests/twister2_plugin_quarantine_test.py::test_if_pytest_handle_quarantine_verify PASSED               [ 42%]
tests/twister2_plugin_quarantine_test.py::test_if_pytest_generate_testplan_with_quarantine PASSED      [ 57%]
tests/quarantine/quarantine_unit_test.py::test_quarantine_load_data PASSED                             [ 71%]
tests/quarantine/quarantine_unit_test.py::test_quarantine_load_data_from_two_files PASSED              [ 85%]
tests/quarantine/quarantine_unit_test.py::test_quarantine_raise_exception_if_data_empty PASSED         [100%]
```

gchwier commented:
Added `skip` mark to tests listed with `--colect-only`, when test should be skipped no matter if because of quarantine or skip keyword in testscace.yml). Also more `--quarantine-list` may be provided. Example usage:
```
$ pytest --zephyr-base=/home/grch/zephyrproject/zephyr --platform=native_posix --testplan-json=twister-out/testplan.json --results-json=twister-out/results.json --quarantine-list=test_quarantine.yml tests/kernel/common tests/kernel/queue --quarantine-list=test_quarantine2.yml --co 
=================================================== test session starts ====================================================
platform linux -- Python 3.8.10, pytest-7.2.0, pluggy-1.0.0
rootdir: /home/grch/zephyrproject/zephyr
plugins: subtests-0.9.0, cov-4.0.0, twister-0.0.1
collected 7 items                                                                                                          

<YamlFile tests/kernel/common/testcase.yaml>
  <YamlTestFunction kernel.common[native_posix]>
  <YamlTestFunction kernel.common.tls[native_posix]>
  <YamlTestFunction kernel.common.misra[native_posix]> skip
  <YamlTestFunction kernel.common.nano32[native_posix]> skip
  <YamlTestFunction kernel.common.nano64[native_posix]> skip
  <YamlTestFunction kernel.common.picolibc[native_posix]> skip
<YamlFile tests/kernel/queue/testcase.yaml>
  <YamlTestFunction kernel.queue[native_posix]>

-------------------- generated testplan file: /home/grch/zephyrproject/zephyr/twister-out/testplan.json --------------------
```

gchwier commented:
After running tests with `--results-json=twister-out/results.json`, filtered tests are listed in json output with "skipped" status and comment in message field:

```
        {
            "suite_name": "tests.kernel.common.testcase.yaml",
            "test_name": "kernel.common.picolibc[native_posix]",
            "nodeid": "tests/kernel/common/testcase.yaml::kernel.common.picolibc[native_posix]",
            "platform": "native_posix",
            "tags": "userspace base picolibc",
            "type": "integration",
            "platform_allow": "",
            "status": "skipped",
            "message": "('/home/grch/upstream_workspace/twister/src/twister2/yaml_test_function.py', 63, 'Skipped: Quarantine: NA')",
            "duration": 0.00010917196050286293,
            "subtests": []
        },

```

fundakol commented on tests/quarantine/quarantine_unit_test.py:
```diff
@@ -0,0 +1,66 @@
+
+def test_quarantine_load_data(tmp_path):
+    quarantine_file = tmp_path / "quarantine.yml"
+    quarantine_file.write_text("""
```
You can use `textwrap.dedent` to have indents in the text (there are some usage in the code already).

gchwier commented on src/twister2/report/test_plan_plugin.py:
```diff
@@ -40,7 +41,7 @@ def __init__(
     def _item_as_dict(self, item: pytest.Item) -> dict:
         """Return test metadata as dictionary."""
-        return dict(
+        test_info = dict(
```
Right, I forgot to check CSV files. I will update it

fundakol commented on src/twister2/report/helper.py:
```diff
@@ -33,6 +33,13 @@ def get_item_type(item: pytest.Item) -> str:
 
+def get_item_skip(item: pytest.Item) -> str:
+    """Return skip reason if test should be skipped."""
+    if marker := item.get_closest_marker('skip'):
```
It may happened there will be more than one `pytest.mark.skip` added to a tests, and this function will return only first occurrence. Same problem will be in `get_item_platform` (no problem with yaml tests, but for python tests user can add more platforms)

gchwier replied:
yes, I agree, if `skip` is used in testcase.yml we can lose info about quarantine, but skip in testcase is obsolete, I found it only in 5 yamls. However, I will try to change it with marking quarantined scenarios in separate plugin.

gchwier commented on src/twister2/filter/quarantine_filter.py:
```diff
@@ -0,0 +1,133 @@
+                logger.debug(f"Skipped test {get_test_name(item)} - quarantine reason: {qelem.comment}")
+                # only mark test to be skipped, this test still will be listed in test plan,
+                # but will not be executed
+                item.add_marker(pytest.mark.skip(f'Quarantine: {qelem.comment}'))
```
It works as expected: test is only marked to be skipped, but hook returns False, it means that test is not deselected. But after   offline discussion I agree that it would be better to not use filter plugin to mark tests to be under quarantine, but to use separate plugin uses `pytest_collection_modifyitems` to mark quarantined tests. I will try to implement such a solution.

gchwier commented:
Next iteration:
Quarantine works as a separate plugin, not under FilterPlugin.
Quarantined tests receive marker `quarantine` and are skipped.
This marker is later used to prepare testplan and test results outputs.
Removed `--quarantine-verify` option. To verify quarantined scenarios, one can use marker as a filter: `-m quarantine` then only quarantined tests are collected and not marked to be skipped.
Removed 'skip' word from --collect-only output (added in previous commit), we can think about adding separate command to list-only collected tests.
Updated/added unit tests.

gchwier commented:
Fixes #23

PerMac commented:
`fixes #23` has to be a part of the commit msg or the PR description to link automatic closing of the issue. Please amend the commit msg and then edit the PRs description to match the msg.

fundakol commented on src/twister2/quarantine_plugin.py:
```diff
@@ -0,0 +1,123 @@
+
+
+@dataclass
+class QuarantineElem:
```
We should not use `abbreviations` in class name. `QuarantineElement` is better name.

fundakol commented on src/twister2/quarantine_plugin.py:
```diff
@@ -0,0 +1,123 @@
+class QuarantineData:
+    qlist: list[QuarantineElement] = field(default_factory=list)
+
+    def __post_init__(self):
```
This will not work if you pass list of `QuarantineElement` instead of list of dicts like so:
```python
qd = QuarantineData([QuarantineElement])
```
Which is what I would do seeing this class. 

I suggest two solution:
- check in `__post_init__` if self.qlist is a list of dict
- add new class method method e.g. `QuarantineData.from_row_data()` which return QuarantuneData instance

gchwier replied:
updated __post_init__ method. Please check it now. Thanks!

PerMac commented on src/twister2/filter/quarantine_filter.py:
```diff
@@ -0,0 +1,125 @@
+            self.architectures = []
+        if not any([self.scenarios, self.platforms, self.architectures]):
+            raise TwisterConfigurationException("At least one of filters ('scenarios', 'platforms', "
+                                                "'architectures') must be specified")
```
the added comment helps with the understanding.
